File: cpp/trapping-rainwater.cpp
// Original title: 42. Trapping Rain Water
/*
Given `n` non-negative integers representing an elevation map
where the width of each bar is `1`, compute how much water
it can trap after raining.
*/
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <cassert>

int find_index_biggest(std::vector<int> group){
  int hold = 0;
  int index = 0;
  for(auto head=group.begin(); head!=group.end(); head++){
    if (hold < *head){
      hold = *head;
      index = head - group.begin();
    }
  }
  return index;
}

int trap(std::vector<int> rain) {
  if (rain.size() <= 2){
    return 0;
  }
  int index_first = find_index_biggest(rain);
  auto clip = rain;
  clip.erase(clip.begin() + index_first);
  int index_second = find_index_biggest(clip);
  if (index_second >= index_first){
    index_second += 1;
  }
  
  int index_left = index_first;
  int index_right = index_second;
  if (index_first > index_second){
    index_left = index_second;
    index_right = index_first;
  }
  auto left = rain.begin() + index_left;
  auto right = rain.begin() + index_right;
  int surface = rain[index_second];
  int already = 0;
  for (auto head = left + 1; head != right; head++){
    int step = surface - *head;
    if (step >= 0) {
      already += step;
    }
  }
  auto subrain = rain;
  auto subleft = subrain.begin() + index_left;
  auto subright = subrain.begin() + index_right;
  subrain.erase(subleft, subright);
  subrain[index_left] = surface;
  int result = trap(subrain) + already;
  return result;
}

int main()
{
  std::vector<int> v = {0,1,0,2,1,0,1,3,2,1,2,1};
  std::cout << trap(v) << std::endl;
  // "6"
}
// Accepted June 29, 2021.

File: leetcode/py-hard/trapping-rainwater.py
#! /usr/bin/env python3

# 42. Trapping Rain Water [hard]
'''
Given `size` non-negative integers representing an elevation map
`rain` where the width of each bar is `1`, compute how much water
it can trap after raining.
   `0 <= size <= 3 * 10^4`
   `0 <= rain[i] <= 10^5`
'''
# Accepted August 14, 2021

import typing as TYPE

def main():
   rain = [0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1] # 6
   solution = Solution()
   number = solution.count_trap(rain)
   print("There are", number, "drops.")
   # 6

class Solution:

   def count_trap(self, rain: TYPE.List[int]) -> int:
      return self.count_left(rain, 0, len(rain) - 1)

   def count_left(self,
      rain: TYPE.List[int],
      start: int,
      stop: int
   ) -> int:
      if (stop - start <= 1):
         return 0
      left = self.probe_left(rain, start)
      if (left >= len(rain)):
         return self.count_right(rain, start, stop)
      water_left = sum([max(0, rain[start] - value) for value in rain[start:left+1]])
      water_right = self.count_left(rain, left, stop)
      return water_left + water_right

   def count_right(self,
      rain: TYPE.List[int],
      start: int,
      stop: int
   ) -> int:
      if (stop - start <= 1):
         return 0
      right = self.probe_right(rain, stop)
      if (right < 0):
         return self.count_left(rain, start, stop)
      water_right = sum([max(0, rain[stop] - value) for value in rain[right:stop+1]])
      water_left = self.count_right(rain, start, right)
      return water_left + water_right

   def probe_right(self, rain: TYPE.List[int], start: int) -> int:
      limit = rain[start]
      head = start - 1
      while (head >= 0):
         if (rain[head] >= limit):
            break
         head -= 1
      return head

   def probe_left(self, rain: TYPE.List[int], start: int) -> int:
      limit = rain[start]
      head = start + 1
      while head < len(rain):
         if (rain[head] >= limit):
            break
         head += 1
      return head
            

# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #

main()